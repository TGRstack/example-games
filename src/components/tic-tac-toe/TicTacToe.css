/* Helpers that should be moved to a central file once used elsewhere  */
.flex_center {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

/* TicTacToe css */
.container {
  min-width: 380px;
  min-height: 420px;
}

.board {
  display: grid;
  grid-template-areas: 'board'; /*  scores */
  grid-template-rows:    repeat(3, 120px);
  grid-template-columns: repeat(3, 1fr);
  grid-column-gap: 10px;
  grid-row-gap: 11px;
  border: 1px solid white;
  box-shadow: darkgrey 2px 2px;
}

.cell {
  background-color: lightgrey;
  color: #000;
}
