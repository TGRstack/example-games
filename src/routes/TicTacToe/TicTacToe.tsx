// tslint:disable max-line-length
import Gx from '@tgrx/gx'
import * as React from 'react'
import Helmet from 'react-helmet'
import {
  Board as TicTacToeBoard,
} from '../../components/tic-tac-toe'
// import * as S from './GameOfLife.css'

function TicTacToe() {
  return (
    <div>
      <Helmet>
        <title>Tic-Tac-Toe</title>
      </Helmet>

      {/* <Gx col={12}>
        <TicTacToeHeader />
      </Gx> */}
      <Gx col={12}>
        <TicTacToeBoard />
      </Gx>
    </div>
  )
}

export default TicTacToe
