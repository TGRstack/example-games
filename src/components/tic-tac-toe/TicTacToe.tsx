import * as React from 'react'
import * as S from './TicTacToe.css'

import { flat } from '../helpers/array'

interface IProps {
  positions?: numbArr[],
}

interface IXYPosition {row: number, col: number}
type numbArr = number[]
type truthyNumber = boolean | number
interface IState {
  winner: number,
  positions: numbArr[],
  player: number,
}

// tslint:disable-next-line no-any
const sumArray = (arr: any[]) => arr.reduce((t: number, n: number) => t += n)

export class Board extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props)

    this.state = {
      player: 1,
      positions: props.positions || [
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
      ],
      winner: 0,
    }
  }

  handlePlayerMove = (e: React.SyntheticEvent, {row, col}: IXYPosition) => {
    e.preventDefault()

    this.setState(s => {
      const currPlayer = s.player + 0

      const positions = [...s.positions]
      positions[row][col] = currPlayer

      const player = currPlayer === 1 ? 2 : 1
      return { positions, player }
      },          () => this.handleNextTurn()
    )
  }

  handleNextTurn() {
    // minimum number of terms is at least 5 (3+2)
    // the sum of the positions grid after 5 turns is 3 + 2*2 = 7
    // the sum of the positions grid after 9 turns is 5 + 4*2 = 13
    const turnPoints = sumArray(flat(this.state.positions))
    const isEnoughTurns = turnPoints >= 7
    const isMaxTurns = turnPoints === 13

    if (!isEnoughTurns) return

    // Enough truns have passed for their to be a winner
    const status = {
      across:    this.checkAcross(),
      // down:      this.checkDown(),
      // diagonal:  this.checkDiagonal(),
    }
    const isWinner = Object.values(status).reduce((win: boolean, n) =>
      win === false
      ? n.winner > 0
      : win
      ,                                           false
    )
    if (isMaxTurns && !isWinner) return this.handleGameTie()
    if (!isWinner) return

    // There's a winner!
    const {winner} = Object.values(status).filter(n => !!n.winner)[0]
    this.handleGameWon(winner)
  }

  handleGameWon(winner: number) {
    alert(`Game Won by Player ${winner}`)
  }
  handleGameTie() {
    alert('Game Tied')
  }
  public checkAcross(debugPositions?: numbArr[]) {
    const {positions} = this.state || {positions: debugPositions}
    const rowsFilled: boolean[] = []
    const rowsPointed: number[] = []

    positions.forEach((row: number[], i: number) => { // [0, 1, 2]
      rowsFilled[i] = row.reduce((t, n) => !!t ? n > 0 : false, true)
      rowsPointed[i] = sumArray(row)
    })

    // only handles 1 row being filled
    const completedRow = rowsFilled.indexOf(true)
    let winner: truthyNumber = false
    switch (rowsPointed[completedRow]) {
      case 3:
        winner = 1
        break
      case 6:
        winner = 2
        break
      default:
        winner = 0
        break
    }

    return {
      winner,
    }
  }

  renderCell(cell: number, dex:  string) {
    const getRowCol = (str: string): IXYPosition => {
      const parts = str.split('_')
      return {
        col: parseInt(parts[1], 10),
        row: parseInt(parts[0], 10),
      }
    }
    const {row, col} = getRowCol(dex)
    return <div
      className={[S.cell, S.flex_center].join(' ')}
      key={dex}
      onClick={e => this.handlePlayerMove(e, {row, col})}
    >
      {cell}
    </div>
  }
  renderBoard() {
    return <div className={[S.board].join(' ')}>
      {this.state.positions.map((rows: number[], row: number) =>
        rows.map((cell: number, col: number) => (
          this.renderCell(cell, `${row}_${col}`)
        )
      ))}
    </div>
  }
  render() {
    return <div className={[S.container].join(' ')}>
      {this.renderBoard()}
    </div>
  }
}
