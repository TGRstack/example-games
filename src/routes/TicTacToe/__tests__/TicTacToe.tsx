/* eslint-disable import/no-extraneous-dependencies */

import { shallow } from 'enzyme'
import * as React from 'react'
import TicTacToe from '../TicTacToe'

describe('<TicTacToe />', () => {
  test('renders', () => {
    const wrapper = shallow(<TicTacToe />)
    expect(wrapper).toMatchSnapshot()
  })
})
