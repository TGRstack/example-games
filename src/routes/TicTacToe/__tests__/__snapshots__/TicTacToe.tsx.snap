// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicTacToe /> renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TicTacToe />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <HelmetWrapper
          defer={true}
          encodeSpecialCharacters={true}
        >
          <title>
            Game of Life
          </title>
        </HelmetWrapper>,
        <Gx
          col={12}
        >
          <Board />
        </Gx>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <title>
            Game of Life
          </title>,
          "defer": true,
          "encodeSpecialCharacters": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Game of Life",
          },
          "ref": null,
          "rendered": "Game of Life",
          "type": "title",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Board />,
          "col": 12,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <HelmetWrapper
            defer={true}
            encodeSpecialCharacters={true}
          >
            <title>
              Game of Life
            </title>
          </HelmetWrapper>,
          <Gx
            col={12}
          >
            <Board />
          </Gx>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <title>
              Game of Life
            </title>,
            "defer": true,
            "encodeSpecialCharacters": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Game of Life",
            },
            "ref": null,
            "rendered": "Game of Life",
            "type": "title",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Board />,
            "col": 12,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
